core/network/src/test/mod.rs

```diff
@@ -664,27 +664,42 @@ pub trait TestNetFactory: Sized {
 			network_sender.clone(),
 			config.clone(),
 			client.clone(),
-			import_queue.clone(),
 			None,
 			tx_pool,
 			specialization,
 		).unwrap();
 
 		let is_offline2 = is_offline.clone();
 		let is_major_syncing2 = is_major_syncing.clone();
+		let import_queue2 = import_queue.clone();
 
 		std::thread::spawn(move || {
 			tokio::run(futures::future::poll_fn(move || {
 				while let Async::Ready(msg) = network_to_protocol_rx.poll().unwrap() {
-					match msg {
-						Some(FromNetworkMsg::PeerConnected(peer_id, debug_msg)) =>
-							protocol.on_peer_connected(peer_id, debug_msg),
-						Some(FromNetworkMsg::PeerDisconnected(peer_id, debug_msg)) =>
-							protocol.on_peer_disconnected(peer_id, debug_msg),
+					let outcome = match msg {
```

Comment:
Here I am also wondering the point of moving the import-queue to here. It seems the code is more complicated, with those `CustomMessageOutcome`, and it also appears somewhat bloated since the `None` case is hit on all branch but one. 

Does performing the operation of sending the message to the import-queue here make `ChainSync` really more deterministic? To me it reads like the logic is just moved around, and made more complicated. 

The changes will also make it much more complicated to break things up into separate tasks. 

I'm in favor of merging `protocol` away from its own thread and into the networking tokio runtime, but I'm less in favor of doing everything in one task. If the async nature of the previous setup caused problems, I'd rather see those problems isolated and fixed, while still allowing the tasks to run separately.

Reply:
That's the code of the tests, which I haven't really updated yet.
At the moment, the tests assume the presence of a background thread and I've had issues with them not being deterministic because of that.
As an example, I'm doing these changes from my personal laptop because the `network` tests don't pass on my office laptop (even with a commit on the master branch from two months ago).

I really would like to make everything deterministic, but I haven't touched the tests code yet.

Reply:
> I'm in favor of merging protocol away from its own thread and into the networking tokio runtime, but I'm less in favor of doing everything in one task.

It's true that everything is one task at the moment, but that's not the ultimate plan.
The general direction of this rework is to move channels as an internal detail of the code, and not expose them through the API.

The future in `service.rs` could still create sub-tasks, but would handle all communications between the "main task" and the sub-tasks internally. It would expose a non-channels API, and all the sub-modlues would expose a non-channels API.

There's actually a commit in #2475 that has handling the protocol as a sub-task, but I removed it afterwards because maintaining that code is quite painful when you're changing API.

Reply:
Ok, I think it makes sense to try to simplify the code, and even running everything in a single task could help de-buggability(since it will all happen one after the other in the same task). 

I'm still wondering why the `import_queue` needs to be moved to the "top-level" right here in the `poll_fn`, why not just keep it in `protocol`? By moving it here you now have to introduce this `outcome`. Does that simplify anything?

Reply:
Ok, that makes sense.